File: backend/services/onboarding/onboarding_service.py
from fastapi import HTTPException
from supabase import Client
from typing import Dict, List, Optional
import logging
from datetime import datetime

logger = logging.getLogger(__name__)

# Onboarding data structure (stored in profile_highlight_values JSONB under '_onboarding' key)
DEFAULT_ONBOARDING = {
    "completed": False,
    "current_step": 0,
    "skipped_sections": [],
    "version": "1.0",
    "last_seen_at": None,
    "mini_tours": {}
}


class OnboardingService:
    """Service for managing onboarding tours using existing profile_highlight_values JSONB column"""
    
    @staticmethod
    def _get_profile_data(user_id: str, client: Client) -> Dict:
        """Helper to get profile_highlight_values JSONB data"""
        result = client.table('creatives').select('profile_highlight_values').eq('user_id', user_id).single().execute()
        
        if not result.data:
            raise HTTPException(status_code=404, detail="Creative profile not found")
        
        # Get existing data or default to empty dict
        return result.data.get('profile_highlight_values') or {}
    
    @staticmethod
    def _update_profile_data(user_id: str, data: Dict, client: Client):
        """Helper to update profile_highlight_values JSONB data"""
        result = client.table('creatives').update({
            'profile_highlight_values': data
        }).eq('user_id', user_id).execute()
        
        if not result.data:
            raise HTTPException(status_code=404, detail="Creative profile not found")
    
    @staticmethod
    async def get_onboarding_status(user_id: str, client: Client) -> Dict:
        """Get the current onboarding status from JSONB column
        
        Returns onboarding data structure with defaults if not exists
        """
        if not client:
            raise ValueError("Authenticated client is required")
        
        try:
            profile_data = OnboardingService._get_profile_data(user_id, client)
            
            # Extract onboarding data (under '_onboarding' key)
            onboarding_data = profile_data.get('_onboarding', DEFAULT_ONBOARDING.copy())
            
            # Ensure all fields exist (backward compatibility)
            for key, default_value in DEFAULT_ONBOARDING.items():
                if key not in onboarding_data:
                    onboarding_data[key] = default_value
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error fetching onboarding status for user_id {user_id}: {str(e)}")
            if 'PGRST116' in str(e) or 'permission denied' in str(e).lower():
                raise HTTPException(status_code=401, detail="Authentication failed")
            raise HTTPException(status_code=500, detail=f"Failed to fetch onboarding status: {str(e)}")

    @staticmethod
    async def update_main_tour_progress(user_id: str, step: int, client: Client) -> Dict:
        """Update current step in main tour"""
        if not client:
            raise ValueError("Authenticated client is required")
        
        try:
            # Get current profile data
            profile_data = OnboardingService._get_profile_data(user_id, client)
            
            # Get or create onboarding data
            onboarding_data = profile_data.get('_onboarding', DEFAULT_ONBOARDING.copy())
            
            # Update step and timestamp
            onboarding_data['current_step'] = step
            onboarding_data['last_seen_at'] = datetime.utcnow().isoformat()
            
            # Write back to JSONB
            profile_data['_onboarding'] = onboarding_data
            OnboardingService._update_profile_data(user_id, profile_data, client)
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error updating progress for user_id {user_id}: {str(e)}")
            raise HTTPException(status_code=500, detail=f"Failed to update progress: {str(e)}")

    @staticmethod
    async def complete_main_tour(user_id: str, client: Client) -> Dict:
        """Mark main tour as completed"""
        if not client:
            raise ValueError("Authenticated client is required")
        
        try:
            profile_data = OnboardingService._get_profile_data(user_id, client)
            onboarding_data = profile_data.get('_onboarding', DEFAULT_ONBOARDING.copy())
            
            onboarding_data['completed'] = True
            onboarding_data['current_step'] = 8
            onboarding_data['last_seen_at'] = datetime.utcnow().isoformat()
            
            profile_data['_onboarding'] = onboarding_data
            OnboardingService._update_profile_data(user_id, profile_data, client)
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error completing tour for user_id {user_id}: {str(e)}")
            raise HTTPException(status_code=500, detail=f"Failed to complete tour: {str(e)}")

    @staticmethod
    async def skip_section(user_id: str, section: str, client: Client) -> Dict:
        """Add section to skipped list"""
        if not client:
            raise ValueError("Authenticated client is required")
        
        valid_sections = ['intro', 'bookings', 'portfolio', 'settings']
        if section not in valid_sections:
            raise HTTPException(status_code=400, detail=f"Invalid section: {section}")
        
        try:
            profile_data = OnboardingService._get_profile_data(user_id, client)
            onboarding_data = profile_data.get('_onboarding', DEFAULT_ONBOARDING.copy())
            
            if section not in onboarding_data['skipped_sections']:
                onboarding_data['skipped_sections'].append(section)
                onboarding_data['last_seen_at'] = datetime.utcnow().isoformat()
                
                profile_data['_onboarding'] = onboarding_data
                OnboardingService._update_profile_data(user_id, profile_data, client)
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error skipping section for user_id {user_id}: {str(e)}")
            raise HTTPException(status_code=500, detail=f"Failed to skip section: {str(e)}")

    @staticmethod
    async def restart_main_tour(user_id: str, client: Client) -> Dict:
        """Reset main tour to beginning"""
        if not client:
            raise ValueError("Authenticated client is required")
        
        try:
            profile_data = OnboardingService._get_profile_data(user_id, client)
            
            # Reset onboarding data
            onboarding_data = DEFAULT_ONBOARDING.copy()
            onboarding_data['last_seen_at'] = datetime.utcnow().isoformat()
            
            profile_data['_onboarding'] = onboarding_data
            OnboardingService._update_profile_data(user_id, profile_data, client)
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error restarting tour for user_id {user_id}: {str(e)}")
            raise HTTPException(status_code=500, detail=f"Failed to restart tour: {str(e)}")

    @staticmethod
    async def complete_mini_tour(user_id: str, tour_id: str, client: Client) -> Dict:
        """Mark mini-tour as completed"""
        if not client:
            raise ValueError("Authenticated client is required")
        
        valid_tours = ['dashboard', 'activity', 'clients', 'public']
        if tour_id not in valid_tours:
            raise HTTPException(status_code=400, detail=f"Invalid tour_id: {tour_id}")
        
        try:
            profile_data = OnboardingService._get_profile_data(user_id, client)
            onboarding_data = profile_data.get('_onboarding', DEFAULT_ONBOARDING.copy())
            
            if 'mini_tours' not in onboarding_data:
                onboarding_data['mini_tours'] = {}
            
            onboarding_data['mini_tours'][tour_id] = True
            
            profile_data['_onboarding'] = onboarding_data
            OnboardingService._update_profile_data(user_id, profile_data, client)
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error completing mini-tour for user_id {user_id}: {str(e)}")
            raise HTTPException(status_code=500, detail=f"Failed to complete mini-tour: {str(e)}")

    @staticmethod
    async def restart_mini_tour(user_id: str, tour_id: str, client: Client) -> Dict:
        """Reset specific mini-tour"""
        if not client:
            raise ValueError("Authenticated client is required")
        
        try:
            profile_data = OnboardingService._get_profile_data(user_id, client)
            onboarding_data = profile_data.get('_onboarding', DEFAULT_ONBOARDING.copy())
            
            if 'mini_tours' not in onboarding_data:
                onboarding_data['mini_tours'] = {}
            
            onboarding_data['mini_tours'][tour_id] = False
            
            profile_data['_onboarding'] = onboarding_data
            OnboardingService._update_profile_data(user_id, profile_data, client)
            
            return onboarding_data
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error restarting mini-tour for user_id {user_id}: {str(e)}")
            raise HTTPException(status_code=500, detail=f"Failed to restart mini-tour: {str(e)}")

File: frontend/src/api/onboardingService.ts
import axios from 'axios';
import { supabase } from '../config/supabase';
import type { 
  OnboardingStatus, 
  UpdateProgressRequest, 
  SkipSectionRequest,
  CompleteMiniTourRequest,
  RestartMiniTourRequest
} from '../types/onboarding';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

// Helper function to check if user is authenticated
const checkAuth = async (): Promise<boolean> => {
  const { data } = await supabase.auth.getSession();
  return !!data.session?.access_token;
};

// Helper function to get auth headers
const getAuthHeaders = async () => {
  const { data } = await supabase.auth.getSession();
  const jwtToken = data.session?.access_token;
  
  return {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    ...(jwtToken && { 'Authorization': `Bearer ${jwtToken}` })
  };
};

/**
 * Get the current onboarding status for the user
 */
export async function getOnboardingStatus(): Promise<OnboardingStatus | null> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    console.log('User not authenticated, skipping onboarding status fetch');
    return null;
  }

  try {
    const headers = await getAuthHeaders();
    const response = await axios.get<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/status`,
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error fetching onboarding status:', error);
    if (error.response?.status === 401) {
      return null;
    }
    throw new Error(error.response?.data?.detail || 'Failed to fetch onboarding status');
  }
}

/**
 * Update the main tour progress
 */
export async function updateMainTourProgress(step: number): Promise<OnboardingStatus> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    throw new Error('User not authenticated');
  }

  try {
    const headers = await getAuthHeaders();
    const body: UpdateProgressRequest = { step };
    const response = await axios.post<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/main-tour/progress`,
      body,
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error updating main tour progress:', error);
    if (error.response?.status === 401) {
      throw new Error('User not authenticated');
    }
    throw new Error(error.response?.data?.detail || 'Failed to update main tour progress');
  }
}

/**
 * Mark the main tour as completed
 */
export async function completeMainTour(): Promise<OnboardingStatus> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    throw new Error('User not authenticated');
  }

  try {
    const headers = await getAuthHeaders();
    const response = await axios.post<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/main-tour/complete`,
      {},
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error completing main tour:', error);
    if (error.response?.status === 401) {
      throw new Error('User not authenticated');
    }
    throw new Error(error.response?.data?.detail || 'Failed to complete main tour');
  }
}

/**
 * Skip a section of the main tour
 */
export async function skipSection(section: string): Promise<OnboardingStatus> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    throw new Error('User not authenticated');
  }

  try {
    const headers = await getAuthHeaders();
    const body: SkipSectionRequest = { section };
    const response = await axios.post<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/main-tour/skip-section`,
      body,
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error skipping section:', error);
    if (error.response?.status === 401) {
      throw new Error('User not authenticated');
    }
    throw new Error(error.response?.data?.detail || 'Failed to skip section');
  }
}

/**
 * Restart the main tour from the beginning
 */
export async function restartMainTour(): Promise<OnboardingStatus> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    throw new Error('User not authenticated');
  }

  try {
    const headers = await getAuthHeaders();
    const response = await axios.post<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/main-tour/restart`,
      {},
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error restarting main tour:', error);
    if (error.response?.status === 401) {
      throw new Error('User not authenticated');
    }
    throw new Error(error.response?.data?.detail || 'Failed to restart main tour');
  }
}

/**
 * Mark a mini-tour as completed
 */
export async function completeMiniTour(tourId: string): Promise<OnboardingStatus> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    throw new Error('User not authenticated');
  }

  try {
    const headers = await getAuthHeaders();
    const body: CompleteMiniTourRequest = { tour_id: tourId };
    const response = await axios.post<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/mini-tour/complete`,
      body,
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error completing mini-tour:', error);
    if (error.response?.status === 401) {
      throw new Error('User not authenticated');
    }
    throw new Error(error.response?.data?.detail || 'Failed to complete mini-tour');
  }
}

/**
 * Restart a specific mini-tour
 */
export async function restartMiniTour(tourId: string): Promise<OnboardingStatus> {
  const isAuthenticated = await checkAuth();
  if (!isAuthenticated) {
    throw new Error('User not authenticated');
  }

  try {
    const headers = await getAuthHeaders();
    const body: RestartMiniTourRequest = { tour_id: tourId };
    const response = await axios.post<OnboardingStatus>(
      `${API_BASE_URL}/api/onboarding/mini-tour/restart`,
      body,
      { headers, withCredentials: true }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error restarting mini-tour:', error);
    if (error.response?.status === 401) {
      throw new Error('User not authenticated');
    }
    throw new Error(error.response?.data?.detail || 'Failed to restart mini-tour');
  }
}
